### ISSUE TITLE
Every non-standard `grpo_variant` drops the KL penalty: `beta` is validated but never used, so `grpo_beta` and the reward-hack controller's β do nothing

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#1232]
Found in a code review; reproduced on `main` at `99d553f5` with trl 0.29.1 (CPU; kernel plus the real GRPO trainer on `hf-internal-testing/tiny-random-gpt2`).

## What happens

The kernel, with a reference policy 3 nats away on every token, `beta` swept from 0 to 100:

```python
import torch

from soup_cli.utils.grpo_variants import apply_variant_loss

torch.manual_seed(0)
logp = torch.randn(2, 4)
kw = dict(logp_new=logp, logp_old=logp - 0.1, advantages=torch.tensor([1.0, -0.5]),
          completion_mask=torch.ones(2, 4))
reference = logp - 3.0  # a reference policy 3 nats away on every token
for variant in ("gspo", "dapo", "dr_grpo", "bnpo", "two_sided", "rft"):
    extra = {"delta": 0.2} if variant == "two_sided" else {}
    losses = [apply_variant_loss(variant, beta=b, reference_logp=reference, **kw, **extra).item()
              for b in (0.0, 0.1, 5.0, 100.0)]
    print(f"{variant:9s}", [round(x, 6) for x in losses])
```

```
gspo      [-0.276293, -0.276293, -0.276293, -0.276293]
dapo      [-0.276293, -0.276293, -0.276293, -0.276293]
dr_grpo   [-1.105171, -1.105171, -1.105171, -1.105171]
bnpo      [-0.276293, -0.276293, -0.276293, -0.276293]
two_sided [-0.276293, -0.276293, -0.276293, -0.276293]
rft       [0.090698, 0.090698, 0.090698, 0.090698]
```

For scale, trl's per-token KL estimator on this reference is 2.0498, so `beta=0.1` should add 0.2050 to the loss and `beta=5` should add 10.2489. The loss is the same to the last digit.

On the real trainer: `GRPOTrainerWrapper.setup()` with `grpo_variant: dapo`, `grpo_beta: 0.1`, `reward_fn: format`, two generations per prompt. trl generates one real batch, including `ref_per_token_logps`, which it computes because `beta != 0`. The LoRA B weights are nudged so the policy differs from the reference (the base with the adapter disabled), and advantages are set to `[1, -1]`. Then `beta` is written to both `trainer.beta` and `trainer.args.beta`, the way the reward-hack controller writes it, and Soup's variant loss is compared with trl's stock GRPO loss on the same inputs:

```
trainer class: _GRPOTrainerVariant_dapo | stock base: GRPOTrainer | trl beta at build: 0.1
ref_per_token_logps in the batch: True | completion tokens: (2, 64)

   beta   Soup dapo loss     |grad|   trl stock loss     |grad|
    0.1         0.000000   0.159877         0.000216   0.040106
    1.0         0.000000   0.159877         0.002159   0.041876
   10.0         0.000000   0.159877         0.021593   0.088595
```

trl's loss moves with `beta` (its KL term). Soup's loss and gradient do not move at all. The variant loss prints as 0 because the two advantages are symmetric (+1 / -1) over two equal-length completions; its gradient is non-zero and identical at every `beta`.

<details><summary>Script for the real-trainer table (run as a file; tiny model from the Hugging Face cache)</summary>

```python
import os
import pathlib
import tempfile

import torch

from soup_cli.config.loader import load_config_from_string
from soup_cli.trainer.grpo import GRPOTrainerWrapper

HERE = pathlib.Path(__file__).parent
with tempfile.TemporaryDirectory(dir=HERE, ignore_cleanup_errors=True) as tmp:
    os.chdir(tmp)
    cfg = load_config_from_string(
        "base: hf-internal-testing/tiny-random-gpt2\n"
        "task: grpo\n"
        "data:\n  train: x.jsonl\n  max_length: 64\n  chat_template: chatml\n"
        "training:\n  epochs: 1\n  batch_size: 2\n  num_generations: 2\n"
        "  reward_fn: format\n  grpo_variant: dapo\n  grpo_beta: 0.1\n  quantization: none\n"
        "output: ./out\n"
    )
    rows = [{"messages": [{"role": "user", "content": f"Think, then say {i}."}]} for i in range(4)]
    wrapper = GRPOTrainerWrapper(cfg, device="cpu")
    wrapper.setup({"train": rows})
    trainer = wrapper.trainer
    stock_cls = type(trainer).__mro__[1]
    print("trainer class:", type(trainer).__name__, "| stock base:", stock_cls.__name__,
          "| trl beta at build:", trainer.beta)
    model = trainer.model
    torch.manual_seed(0)
    batch = next(iter(trainer.get_train_dataloader()))
    inputs = trainer._prepare_inputs(batch)
    print("ref_per_token_logps in the batch:", inputs.get("ref_per_token_logps") is not None,
          "| completion tokens:", tuple(inputs["completion_ids"].shape))
    n = inputs["advantages"].shape[0]
    inputs["advantages"] = torch.tensor([1.0, -1.0] * (n // 2))  # a group with a winner
    with torch.no_grad():  # move the policy away from the reference (adapters off = reference)
        g = torch.Generator().manual_seed(1)
        for name, p in model.named_parameters():
            if "lora_B" in name:
                p.add_(0.5 * torch.randn(p.shape, generator=g))
    model.train()
    trainable = [p for p in model.parameters() if p.requires_grad]

    def loss_and_grad(fn):
        model.zero_grad(set_to_none=True)
        torch.manual_seed(3)
        loss = fn()
        loss.backward()
        gn = torch.sqrt(sum((p.grad ** 2).sum() for p in trainable if p.grad is not None))
        return loss.item(), gn.item()

    print(f"\n{'beta':>7s} {'Soup dapo loss':>16s} {'|grad|':>10s} {'trl stock loss':>16s} {'|grad|':>10s}")
    for b in (0.1, 1.0, 10.0):
        trainer.beta = b  # what reward_hack_control._apply_coefficient writes ...
        trainer.args.beta = b  # ... to both places
        v_loss, v_grad = loss_and_grad(lambda: trainer._compute_loss(model, inputs))
        s_loss, s_grad = loss_and_grad(lambda: stock_cls._compute_loss(trainer, model, inputs))
        print(f"{b:7.1f} {v_loss:16.6f} {v_grad:10.6f} {s_loss:16.6f} {s_grad:10.6f}")
    os.chdir(HERE)
```

</details>

## Why

- `src/soup_cli/utils/grpo_variants.py:252-258` turns `beta` into `beta_f` and range-checks it, and nothing after that reads `beta_f`. `reference_logp` appears only in the signature (`:196`) and the docstring (`:214-215`, "used by `standard`-with-beta and `dr_grpo`").
- The variant trainer passes both in (`src/soup_cli/trainer/grpo.py:130-149`), and its `_compute_loss` override (`grpo.py:165-171`) replaces trl's `_compute_loss`. trl's method is where `beta * per_token_kl` is added (`trl/trainer/grpo_trainer.py:2182-2222`), so replacing it drops the KL term.
- trl still runs the reference forward on every step while `beta != 0` (`grpo_trainer.py:1836`), so the run pays for a reference it never uses.
- `reward_hack_mitigation: kl_control` / `pid_lagrangian` actuate β through `_apply_coefficient` (`src/soup_cli/utils/reward_hack_control.py:993-1017`), which dual-writes `args.beta` because "Soup's `_GRPOTrainerVariant` reads `self.args.beta`". It reads it and discards it, so under any non-standard variant the controller's actuation is inert. That is separate from the controller issues tracked in #371.

## Fix path

This needs a per-variant decision. Some published variants drop the KL term on purpose (DAPO removes it), while others keep it.

- **Wire it** where a variant should have KL: add trl's estimator (`exp(ref - logp) - (ref - logp) - 1`), masked, and aggregated the way that variant aggregates its policy term (token mean for `dapo` / `bnpo` / `two_sided`, sequence level for `gspo`, sum over tokens for `dr_grpo`; decide explicitly for `rft`). Raise when `beta > 0` and `reference_logp` is missing, rather than skipping silently.
- **Refuse it** where a variant is KL-free by design: reject a non-default `grpo_beta` and `reward_hack_mitigation: kl_control|pid_lagrangian` at config parse, with a message naming the variant, and pass `beta=0` to trl so the reference forward is not run.

## Acceptance

- [ ] Kernel: for each variant with KL, `loss(beta=1) - loss(beta=0)` equals the masked KL estimate (within 1e-6), and `beta > 0` with `reference_logp=None` raises.
- [ ] Real trainer (tiny CPU model, as above): the variant's loss and gradient respond to `beta` exactly as trl's stock loss does on the same inputs, or the config is refused at parse.
- [ ] The reward-hack controller in `kl_control` mode changes the loss under a non-standard variant, or the combination is refused with a message naming the variant.
- [ ] A KL-free variant runs with trl's `beta == 0`, so no reference forward happens.
- [ ] The `apply_variant_loss` docstring matches what each variant does.
